salt/utils/etcd_util.py

```diff
@@ -497,11 +532,206 @@ def tree(self, path):
         return ret
 
 
+class EtcdApiV3Adapter(EtcdClient):
+    """
+    Since etcd3 has no concept of directories, this class leaves some methods unimplemented.
+
+    These are: ls and write_directory.
```

Comment:
I suppose we could pretend, although your suggestion wouldn't quite work I don't think -- it would also show the values of subdirectories since results would hold _anything_ that started with `path`.  I guess it depends on if we want to force directories upon v3 even when the underlying service doesn't have them.  The decision to not support them was based solely on the intuition that users on v3 would already know that v3 doesn't have directories and the concept of `ls` and `write_directory` doesn't make sense with that in mind.  If you think we should pretend, I know how to implement it and I don't think it would take very long.

Reply:
I guess I'm conflicted then 😂 

I *think* that in some things like say cache, we provide directory-like APIs even though directories don't actually exist -- because the localfs is built on the filesystem and that mechanism ends out being exposed. So we just fake it where it doesn't exist.

Maybe that *only* makes sense for the cache and not generally for etcd support. Thoughts?

Reply:
As we discussed - yes is forever, no is temporary.

I think it's fine for us to say that *currently* we don't need `write_directory` (I'm also OK on punting `ls`). There are a couple of weird edge cases that write_directory would produce that we don't really want to implement right now. We should just document those to the best of our current knowledge and then leave it for future us or someone else to take on when they need that functionality.